### PR TITLE
Resolves issues with flag capping in BG's like AB

### PR DESCRIPTION
**Pull request by BenDol**
## 🍰 Pullrequest
Resolves https://github.com/cmangos/issues/issues/1250 

The solution here is that we need to send a player update for the objects state will sometimes not sync correctly with the client since the object is invisible when the state change takes place.
In order to maintain its object state on the clients we will keep track.

### Proof
https://github.com/cmangos/issues/issues/1250 

### Issues
https://github.com/cmangos/issues/issues/1250 

### How2Test
* login 2 chars of opposite factions
* .debug bg
* join Arathi Basin on each character
* cap the same flag post on each character
* do it forever because its fixed

### Todo / Checklist

### CODE REVIEW

File: src/game/BattleGround/BattleGround.h [hunk 1]
```diff
@@ -616,6 +619,8 @@ class BattleGround
         float m_TeamStartLocZ[PVP_TEAM_COUNT];
         float m_TeamStartLocO[PVP_TEAM_COUNT];
         float m_startMaxDist;
+
+        GuidVector m_objUpdates;
```
Comment (killerwife) [inline src/game/BattleGround/BattleGround.h]: Why not use std::set instead. You require uniqueness, so that solves it for you implicitly.

**Comment by killerwife** [comment]
Damn, I would love to investigate this and look for it in a sniff, but I sure as hell do not have time for such a rabbit hole. So the issue is that we do not send some data, which causes the data to be stale in the client and then client thinks that a certain value is still set to the old one? The true fix to me would be to properly send that all on the object showing up again, but I digress. @Warlockbugs your opinion?

**Comment by BenDol** [comment]
Yeah its a strange issue, if anything I can highlight it here if someone wants to find a more robust solution to the issue. I tried to dig in and see where it was going wrong with little success. But my time is spreading thin haha.

File: src/game/BattleGround/BattleGround.h (hunk 1 above)
Comment (BenDol) [reply src/game/BattleGround/BattleGround.h]: Yeah good point will update that

**Comment by killerwife** [comment]
@Shauren does this in any way ring a bell for anything in TrinityCore?

**Comment by Shauren** [comment]
Yes, this is the same issue as .gobject move/turn commands had (had, but the fix for those is not applicable here)

Basically, client stores recently deleted (OutOfRange update) objects and brings them back using old state when new CreateObject is received, ignoring new state sent in that CreateObject packet

commands work around it by simply spawning a new object with new GUID

**Comment by killerwife** [comment]
So the "true" fix is to have dynamic guids, and this is a bandaid until we have that like TC does, correct?

**Comment by Shauren** [comment]
No, as I said at the beginning - the fix is not applicable to this BG flag case (guid is not supposed to change for the same object), treat what I said as simply explaining the cause

That client object cache behavior is not really well handled/reversed

**Comment by BenDol** [comment]
Well the ball is rolling for this fix if someone wants to implement it correctly

**Comment by Yatzii93** [comment]
I tried this out and for me it didn't resolve the issue....hmmmm

**Comment by Lymphome** [comment]
Hi, don't know if it is the good place to post that (my first post on github) but following the comments from Kelno I modified SpawnBGObject() in BattleGround.cpp to solve this flag problem :

if (respawntime == 0)
{
        obj->Respawn(); // respawn => SetRespawnTime(0);
        if (obj->GetGoType() == GAMEOBJECT_TYPE_BUTTON)
                obj->SetInUse(false);   // GO_STATE_READY

        if (obj->GetLootState() == GO_JUST_DEACTIVATED)
                obj->SetLootState(GO_READY);

} else {
        // flags are buttons (except neutral flags) : need to force state update to clients
        if (obj->GetGoType() == GAMEOBJECT_TYPE_BUTTON)
        {
                obj->SetInUse(true);    // GO_STATE_ACTIVE
                obj->SendForcedObjectUpdate();
        }

        obj->SetLootState(GO_JUST_DEACTIVATED);
        obj->SetRespawnDelay(respawntime);
        obj->SetForcedDespawn();
}

I don't know if it is the good way but it works

**Comment by killerwife** [comment]
Please, if you can, re-edit your code for it to be fully visible.

**Comment by Yatzii93** [comment]
Interesting -
These are custom right ?
> obj->SetRespawnDelay(respawntime);
> obj->SetForcedDespawn();

**Comment by Lymphome** [comment]
These parts come from Entities/[Creature|GameObject].cpp, there are already in. I add nothing else but this small part of code. I did a lots of tries to finally make it work. The order of the calls are important : you have to change states and respawntimes when the object is visible in the client (as Kelno notices it).

**Comment by Lymphome** [comment]
AV quickly tested, it seems ok. Just have to change ally contested flag type for cim : put 'flags' field to 0 like other cim flags (DB Gob template id 179287). Tower flags are ok. May have other things to test...

**Comment by killerwife** [comment]
Why would you need to change EY banner types?

**Comment by Lymphome** [comment]
With type = 26, they are all displayed and don't despawn with this patch. With type 24 it works and this type seems more appropriate to the object (flag drop against flag stand)

**Comment by Yatzii93** [comment]
@Lymphome Can you do a seperate pull request so I can see what you've done in an easy format to achieve the results?

**Comment by cyberium** [comment]
Does not fix the issue. A proper way will be to give another guid to the object. Hopfully something is comming ;)

**Comment by killerwife** [comment]
Soon™ 
image

**Pull request closed without merging**